holoviews/core/spaces.py

```diff
@@ -496,11 +506,17 @@ def _initial_key(self):
         values on the key dimensions.
         """
         key = []
+        undefined = []
         for kdim in self.kdims:
             if kdim.values:
                 key.append(kdim.values[0])
             elif kdim.range:
                 key.append(kdim.range[0])
+            else:
+                undefined.append(kdim)
+        if undefined:
+            raise KeyError('dimensions do not specify a range or values, '
+                           'cannot supply initial key' % ', '.join(undefined))
```

Comment:
No harm having this check here but I was thinking that ``DynamicMap`` should also be doing similar (if not identical) validation. I definitely want to be checking that kdims have the appropriate range settings there.